### PR TITLE
Cross compile MainLoop

### PR DESCRIPTION
**Pull request by kovalromank**
Add support for cross compiling the project. Created a custom cmake toolchain that uses a raspberry pi provided gcc, g++ and other tools/libraries.

The raspberry pi tools didn't support the `stdc++fs` library that we link in the root CMakeLists.txt so I had to remove the following line:

```cmake
target_link_libraries(MainLoopLib stdc++fs)
```

This made it so we can't use `<experimental/filesystem>` , which the Logger currently uses. I'll still try and see if I can include that library but in the meantime we can cross compile the project without the Logger.

To cross compile you will need the raspberry pi tools. Clone the `https://github.com/raspberrypi/tools` repository somewhere, the final size of the folder is about 1gb so it takes some time.
Then set the `RPI_TOOLS` environment variable to the path of the repository you just cloned and run the `build-arm.sh` script.
Example:
```bash
RPI_TOOLS=/path/to/rpi-tools ./build-arm.sh
```

The built executable will be at `/build-arm/MainLoop`.

Next steps is so try and include the experimental/filesystem library or modify the Logger to not use it.

### CODE REVIEW

**Comment by jonapap** [comment]
Do the raspberry pi tools support boost filesystem (https://www.boost.org/doc/libs/1_39_0/libs/filesystem/doc/index.htm)? If so, we could switch to this.

**Comment by kovalromank** [comment]
Cross compilation works. The project now depends on boost's filesystem to install it on Ubuntu/Debian run `apt-get install libboost-filesystem-dev` or `apt-get install libboost-all-dev`.

To cross compile you still need to set two environment variables, the `RPI_TOOLS` as mentioned above and `BOOST_DIRECTORY`. 

`BOOST_DIRECTORY` needs to be a path pointing to the location of compiled boost libraries for arm. I created a script that does this for you, run the `build-boost-arm.sh` file somewhere and set the environment variable to the created directory, right now it would be `boost_1_70_0`. For example:

```bash
# cd out of the repository
cd ..

# this will create a boost_1_70_0 directory
./rocket-code-2020/build-boost-arm.sh

export BOOST_DIRECTORY=/path/to/boost_1_70_0

# compile for arm
cd ./rocket-code-2020
./build-arm.sh
```

**Comment by ajayyy** [comment]
Can you update https://github.com/uorocketry/rocket-code-2020/wiki/Getting-started, and maybe add the info from your comment to a wiki page?

**Comment by kovalromank** [comment]
Done, [Cross compiling for the Raspberry Pi](https://github.com/uorocketry/rocket-code-2020/wiki/Cross-compiling-for-the-Raspberry-Pi).

**Comment by jonapap** [comment]
End-to-end test seems to work fine on the Pi (this is using the binary compiled on my desktop and transferred to the Pi): 
image

**Comment by jonapap** [comment]
However, these changes prevent us from building locally on the Pi. The Boost library on the Pi is version 1.67, but we require 1.70. Can we lower the version of the library to 1.67?
```
pi@raspberrypi:~/rocket-code-2020 $ ./build.sh 
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost: Found unsuitable version "1.67.0", but required is at
  least "1.70.0" (found /usr/include)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:391 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:100 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/pi/rocket-code-2020/build/CMakeFiles/CMakeOutput.log".
See also "/home/pi/rocket-code-2020/build/CMakeFiles/CMakeError.log".
```

**Review by ajayyy — APPROVED** [review]

**Comment by jonapap** [comment]
Still not working. I'm trying to figure out what is happening right now. Boost is installed.
```
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Boost: /usr/include (found suitable version "1.67.0", minimum required is "1.67.0") found components: system filesystem 
-- Configuring done
CMake Error at CMakeLists.txt:103 (add_dependencies):
  The dependency target "/usr/lib/arm-linux-gnueabihf/libboost_filesystem.so"
  of target "MainLoopLib" does not exist.


CMake Error at CMakeLists.txt:103 (add_dependencies):
  The dependency target "/usr/lib/arm-linux-gnueabihf/libboost_system.so" of
  target "MainLoopLib" does not exist.


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.

```

**Comment by kovalromank** [comment]
Are you running build.sh or build-arm.sh on the raspberry pi?

**Comment by jonapap** [comment]
`build.sh`. Isn't `build-arm.sh` for cross-compile?

**Review by jonapap — APPROVED** [review]

File: build-and-run.sh [hunk 1]
```diff
@@ -1,3 +1,5 @@
 #!/bin/bash
 
+set -euf -o pipefail
+
```
Comment (jonapap) [inline build-and-run.sh:4]: While I partly agree with having this here, this should be in a separate PR.
```suggestion
```

Comment (jonapap) [reply build-and-run.sh:4]: Also, the `-u` flag causes issues if we don't specify all the variables, which is something that the script should handle fine.

Comment (kovalromank) [reply build-and-run.sh]: Yeah I didn't realize I staged build.sh and build-and-run.sh, changed them back to what they were previously.

**Review by ajayyy — APPROVED** [review]

**Review by jonapap — DISMISSED** [review]

**Review by ajayyy — APPROVED** [review]
